src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -76,30 +79,39 @@ public BoundLambda(SyntaxNode syntax, UnboundLambda unboundLambda, BoundBlock bo
             );
         }
 
-        public TypeWithAnnotations GetInferredReturnType(ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        public TypeWithAnnotations GetInferredReturnType(ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
         {
             // Nullability (and conversions) are ignored.
-            return GetInferredReturnType(conversions: null, nullableState: null, ref useSiteDiagnostics);
+            return GetInferredReturnType(conversions: null, nullableState: null, ref useSiteInfo);
         }
 
         /// <summary>
         /// Infer return type. If `nullableState` is non-null, nullability is also inferred and `NullableWalker.Analyze`
         /// uses that state to set the inferred nullability of variables in the enclosing scope. `conversions` is
         /// only needed when nullability is inferred.
         /// </summary>
-        public TypeWithAnnotations GetInferredReturnType(ConversionsBase conversions, NullableWalker.VariableState nullableState, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        public TypeWithAnnotations GetInferredReturnType(ConversionsBase conversions, NullableWalker.VariableState nullableState, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
         {
             if (!InferredReturnType.UseSiteDiagnostics.IsEmpty)
             {
-                if (useSiteDiagnostics == null)
+                if (useSiteInfo.Diagnostics == null)
                 {
-                    useSiteDiagnostics = new HashSet<DiagnosticInfo>();
+                    useSiteInfo.Diagnostics = new HashSet<DiagnosticInfo>();
```

Comment:
>useSiteInfo.Diagnostics = new HashSet<DiagnosticInfo>(); [](start = 20, length = 56)

This feels like the type of detail that should be encapsulated by CompoundUseSiteInfo, not exposed publicly.

Reply:
> This feels like the type of detail that should be encapsulated by CompoundUseSiteInfo, not exposed publicly.

Allocating the hashsets outside of the ```CompoundUseSiteInfo<AssemblySymbol>``` type doesn't violate any design behind the type, it is mutable. However, I will adjust this code to call helpers that the ```CompoundUseSiteInfo<AssemblySymbol>``` already provides, they handle allocations as well.

---
In reply to: [361725926](https://github.com/dotnet/roslyn/pull/40560#discussion_r361725926) [](ancestors = 361725926)

Reply:
Adjusted in [https://github.com/dotnet/roslyn/pull/40640](https://github.com/dotnet/roslyn/pull/40640)

---
In reply to: [361734870](https://github.com/dotnet/roslyn/pull/40560#discussion_r361734870) [](ancestors = 361734870,361725926)